tiledb/common/memory_tracker.h

```diff
@@ -179,10 +180,14 @@ class MemoryTrackerResource : public tdb::pmr::memory_resource {
   explicit MemoryTrackerResource(
       tdb::pmr::memory_resource* upstream,
       std::atomic<uint64_t>& total_counter,
-      std::atomic<uint64_t>& type_counter)
+      std::atomic<uint64_t>& type_counter,
+      const std::atomic<uint64_t>& memory_budget,
+      std::function<void()> on_budget_exceeded)
```

Comment:
We have been using templated arguments to represent callbacks, but I can't find a way to make this class generic to accomodate this, so went with an `std::function<void()>`.